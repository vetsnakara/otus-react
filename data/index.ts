import type { TaskListItemData } from 'types'

export const getTasks = ({
  titles,
  completed = false
}: {
  titles: string[]
  completed?: boolean
}): Array<TaskListItemData> => {
  return titles.map((title) => ({
    id: title,
    completed,
    title
  }))
}
