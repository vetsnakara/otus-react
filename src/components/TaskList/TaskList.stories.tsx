import React from 'react'
import { actions } from '@storybook/addon-actions'

import { getTasks } from '../../../internals/testUtils'

import TaskList from './TaskList'

export default {
  title: 'TaskList',
  component: TaskList,
  decorators: [(story) => <div className="container">{story()}</div>]
}

const taskListActions = actions({
  onToggleComplete: 'Toggle task',
  onRemove: 'Remove task'
})

export const NoTasks: React.FC<{}> = () => <TaskList {...taskListActions} />

export const WithTasks: React.FC<{}> = () => (
  <TaskList
    tasks={getTasks({ titles: ['One', 'Two', 'Three'] })}
    {...taskListActions}
  />
)

export const WithCompletedTasks: React.FC<{}> = () => (
  <TaskList
    tasks={[
      ...getTasks({ titles: ['One', 'Two'] }),
      ...getTasks({ titles: ['Three', 'Four'], completed: true })
    ]}
    {...taskListActions}
  />
)
