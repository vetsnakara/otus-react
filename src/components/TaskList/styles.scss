.task-list {
  margin: 0;
  padding: 0;
  list-style-type: none;

  &__item:not(:last-child) {
    margin-bottom: 5px;
  }
}