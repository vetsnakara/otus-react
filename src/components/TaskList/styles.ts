import styled from '@emotion/styled'

/**
 * TaskListContainer
 */
export const TaskListContainer = styled.ul`
  margin: 0;
  padding: 0;
  list-style-type: none;
`

TaskListContainer.displayName = 'TaskListContainer'

/**
 * TaskListItem
 */
export const TaskListItem = styled.li`
  &:not(:last-child) {
    margin-bottom: 5px;
  }
`

TaskListItem.displayName = 'TaskListItem'
