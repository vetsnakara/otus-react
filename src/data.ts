import type { TaskListItemData } from '../src/interfaces'

export const getTasks = ({
  titles,
  completed = false
}: {
  titles: string[]
  completed?: boolean
}): Array<TaskListItemData> => {
  return titles.map((title) => ({
    id: title,
    completed,
    title
  }))
}
