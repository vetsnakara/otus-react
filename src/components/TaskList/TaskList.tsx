import React, { FC } from 'react'

import { Task } from '../Task'

import { TaskListProps } from 'types'

import './styles.scss'

export const TaskList: FC<TaskListProps> = ({
  tasks = [],
  onToggleComplete,
  onRemove
}) => {
  if (tasks.length === 0) return null

  return (
    <ul className="task-list" data-test="task-list">
      {tasks.map((task) => {
        const { id } = task

        return (
          <li key={id} className="task-list__item" data-test="task-list-item">
            <Task
              task={task}
              onToggleComplete={() => onToggleComplete(id)}
              onRemove={() => onRemove(id)}
            />
          </li>
        )
      })}
    </ul>
  )
}
