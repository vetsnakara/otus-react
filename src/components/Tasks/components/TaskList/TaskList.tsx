import React, { FC } from 'react'

import { Task } from './components/Task'

import { TaskListProps } from 'types'

import { TaskListContainer as Container, TaskListItem as Item } from './styles'

export const TaskList: FC<TaskListProps> = ({
  tasks = [],
  onToggleComplete,
  onRemove
}) => {
  if (tasks.length === 0) return null

  return (
    <Container>
      {tasks.map((task) => {
        const { id } = task

        return (
          <Item key={id}>
            <Task
              task={task}
              onToggleComplete={() => onToggleComplete(id)}
              onRemove={() => onRemove(id)}
            />
          </Item>
        )
      })}
    </Container>
  )
}
