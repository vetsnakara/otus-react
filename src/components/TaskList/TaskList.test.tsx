import React from 'react'
import { shallow, mount } from 'enzyme'

import type { TaskListProps } from 'types'

import { TaskList } from './TaskList'

import { getTasks } from 'data'

interface SetupProps extends Partial<TaskListProps> {
  useMount?: boolean
}

/**
 * TaskList wrapper factory function
 */
const setup = ({
  tasks,
  onToggleComplete = jest.fn(),
  onRemove = jest.fn(),
  useMount = false
}: SetupProps = {}) => {
  const setupProps: TaskListProps = {
    tasks,
    onToggleComplete,
    onRemove
  }

  return useMount
    ? mount(<TaskList {...setupProps} />)
    : shallow(<TaskList {...setupProps} />)
}

describe('TaskList', () => {
  test('Nothing is rendered if no tasks passed', () => {
    const wrapper = setup()
    const taskList = wrapper.find('TaskListContainer')
    expect(taskList.length).toBe(0)
  })

  test('All tasks are rendered correctly', () => {
    const tasks = getTasks({ titles: ['One', 'Two', 'Three'] })

    const onToggleComplete = jest.fn()
    const onRemove = jest.fn()

    const wrapper = setup({ tasks, onToggleComplete, onRemove, useMount: true })

    // item elements are rendered
    const items = wrapper.find('TaskListItem')
    expect(items.length).toBe(tasks.length)
  })

  test('Toggle callback is called properly', () => {
    const tasks = getTasks({ titles: ['One'] })
    const { id: firstTaskId } = tasks[0]

    const onToggleComplete = jest.fn()

    const wrapper = setup({
      useMount: true,
      tasks,
      onToggleComplete
    })

    const toggleBtn = wrapper
      .find('Task')
      .first()
      .find('ToggleButton')
      .childAt(0)

    toggleBtn.simulate('click')

    expect(onToggleComplete).toHaveBeenCalledTimes(1)
    expect(onToggleComplete).toHaveBeenCalledWith(firstTaskId)
  })

  test('Remove callback is called properly', () => {
    const tasks = getTasks({ titles: ['One'] })
    const { id: firstTaskId } = tasks[0]

    const onRemove = jest.fn()

    const wrapper = setup({
      useMount: true,
      tasks,
      onRemove
    })

    const removeBtn = wrapper
      .find('Task')
      .first()
      .find('RemoveButton')
      .childAt(0)

    removeBtn.simulate('click')

    expect(onRemove).toHaveBeenCalledTimes(1)
    expect(onRemove).toHaveBeenCalledWith(firstTaskId)
  })
})
