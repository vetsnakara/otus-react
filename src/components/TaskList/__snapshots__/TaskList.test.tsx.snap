// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList All tasks are rendered correctly 1`] = `[Function]`;

exports[`TaskList Nothing is rendered if no tasks passed 1`] = `[Function]`;

exports[`TaskList Remove callback is called properly 1`] = `[Function]`;

exports[`TaskList Toggle callback is called properly 1`] = `[Function]`;
